### DISCUSSION TITLE
Dashboard Schema Versioning

### DISCUSSION THREAD

**Discussion by cndonovan** [perses/perses#1186] [Design Discussions]
## Context
My company is planning to ship some dashboards based on Perses to customers. Additionally, we know we will introduce breaking changes to the dashboard schema in the future. Therefore, we want a basic plan for what will happen when we introduce breaking changes to the dashboard schema (i.e. we don't want dashboards to break for our customers).  

## Goal
Kick off a discussion for a basic plan for what will happen when we introduce a breaking change to the Perses dashboard schema.

## Proposal
This is a straw-dog proposal. I am writing down a simple plan so that we have something to discuss and compare against. 

## Steps

1. Add a field to represent the version number for the dashboard schema:
```
{
  kind: 'Dashboard',
  version: 1, <-- Additional field. `version` or `schema_version`? 
  metadata: {...},
  spec: {...},
}
```
2. Implement the `DashboardSchemaMigrator`. This is client-side code that is able to migrate from one version to the next version. 
Example: Suppose the latest version is 3. In this case, the code is able to convert a dashboard from version 1 to version 2, and from version 2 to version 3. 
3. Whenever a dashboard is loaded, we run this code client-side. The updates are chained together. 
Example: The dashboard stored on the backend is version 1. The latest version is 3. The migrator converts from version 1 to version 2, then from version 2 to version 3. 
4. Optionally, the client can give users the option to store the migrated dashboard in the backend, so we don't have to keep migrating on each page load.

## Outcome
So, what happens when we make a breaking change?
Whenever we make a breaking change to the dashboard schema, we have to update the dashboard schema migrator. If we do this, existing dashboards won't break. 
Example: The latest version of the dashboard schema is 33. A developer introduces a breaking change. The developer needs to update `DashboardSchemaMigrator` and write code that converts a dashboard from version 33 to version 34. After doing this, the rest of the system works.

**Comment by saminzadeh** [comment]
> Add a field to represent the version number for the dashboard schema:

If we want to be consistent with Kubernetes resources, `apiVersion` might be a good name for this. [More info](https://matthewpalmer.net/kubernetes-app-developer/articles/kubernetes-apiversion-definition-guide.html)


Not to complicate this further but how are we thinking this works for plugins? I suspect most of the breaking changes will occur in the plugin-defined spec rather than the top level dashboard spec. Plugins will be particularly challenging for a few reasons:
- The spec is defined directly in the plugin code which may or may not live with the Perses core source code
- Plugins may need to know which version of Perses they are running in case they start using plugin hooks that are not available in a newer version

Similar to your idea, we may be able to add an `apiVersion` to each Panel resource as well. Then the plugin can define its own`SchemaMigrator` that gets injected by the `DashboardSchemaMigrator` during migration to migrate panels of its type.

Example where Dashboard is apiVersion 1 but PrometheusTimeSeriesQuery is apiVersion 2 
```yaml
kind: Dashboard
apiVersion: 1
spec:
  panels:
    GrpcbyCode:
      kind: Panel
      spec:
        display:
          name: Grpc by Code
        plugin:
          kind: TimeSeriesChart
          spec: {}
        queries:
        - kind: TimeSeriesQuery
          spec:
            plugin:
              kind: PrometheusTimeSeriesQuery
              apiVersion: 2
              spec:
                query: sum(irate(grpc_client_handled_total[1m])) by (grpc_code) 
```

Maybe, the first iteration doesn't need to do this. The main Dashboard migrator can handle first-class plugins (plugins maintained by the core project) directly in its source code to avoid the complexity.

**Reply by sjcobb** [reply]
Yeah I like this idea of narrowing the scope to only worry about migrating first-class plugins initially.

**Reply by sjcobb** [reply]
And ugh the snake_case discussion againn 😂 
@Nexucis how closely are we trying to follow the Kubernetes data model since they use camelCase for [their keys](https://kubernetes.io/docs/reference/kubernetes-api/common-parameters/common-parameters/#propagationPolicy). Do you think we should go with something like `schema_version` or `apiVersion` like Shan suggested? 

I just hope we are consistent and don't end up in a situation where different casing conventions are used in different parts of the schema.

**Comment by Nexucis** [comment]
Reviving a bit this topic as I got some times to think about it.

So yes indeed we will have to introduce the version number on each plugin. With the new Perses architecture we will be able to have multiple version of the same plugin. So that would mitigate the upgrade of a version as a first step.

Then I guess the upgrade can be done through a cuelang script like we are doing with the migration from Grafana to Perses dashboard format.
So things will be done on backend side, not frontend.

**Reply by Nexucis** [reply]
Continuing thinking about this topic:

1. Like mentioned above, we will use cuelang to perform the migration from a version to another. For the same plugin we can have multiple different script to upgrade a plugin. Each script/file will be associated to a range of version.

For example: 

```json
{
  "upgrading01.cue": ">= v0.3.4"
  "upgrading02.cue": "> v0.1.0, < v0.3.4"
}
```

When there is no script to use, then upgrading the plugin will be automatic as we are assuming there is no data-model changes, just the version to upgrade in the data-model.


Performing the migration of the plugin will be done on the server side probably when the server is starting.

2. Perses is today used for some project just for the Dashboard data-model. So it means that he project supporting Perses needs to support the same core of plugins. Which is fine for the list of panel because they are really basics. But once we are introducing the version of the plugin in the data-model, that would mean for this kind of project to follow the same versioning.

A way to avoid that would be to introduce a source where the plugin is coming from. Like Kubernetes with the CRDs, you will have to specify the registry of the plugin.

A rough idea to reflect these ideas would be to update the plugin data-model by introducing a plugin metadata:

```yaml
plugin:
  kind: TimeSeriesChart
  metadata:
    version: "0.4.1"
    registry: "persesdev"
  spec: {}
```
